### ISSUE TITLE
[bug] xml docs generics+`inheritdoc`+`typeparamref`

### ISSUE THREAD

**Issue by ZacharyPatten** [dotnet/roslyn#54069]
There is a bug in the displayed XML documentation regarding generics+`inheritdoc`+`typeparamref` that is currently reproduceable in both Visual Studio and Visual Studio Code.

Consider the following top-level program in .NET 5+
```cs
Test<int> a = new();
var b = a.Clone();

public class Test<T> : ICloneable<Test<T>>
{
	/// <inheritdoc/>
	public Test<T> Clone() => new();
}

/// <summary>A type that has clonable instances.</summary>
/// <typeparam name="T">The type of instances that can be cloned.</typeparam>
public interface ICloneable<T>
{
	/// <summary>Clones a <typeparamref name="T"/>.</summary>
	/// <returns>A clone of the <typeparamref name="T"/>.</returns>
	public T Clone();
}
```

Now hover over the `Clone` method on line 2 where we have `var b = a.Clone();`.

Screenshot from Visual Studio:
![image](https://user-images.githubusercontent.com/3385986/121825073-7d982280-cc7e-11eb-9fa4-d0f7bb79738f.png)

Screenshot from Visual Studio Code:
![image](https://user-images.githubusercontent.com/3385986/121825080-90aaf280-cc7e-11eb-948c-1b6ddbafb749.png)

The "T" shown in the pop-up XML comments is not correct. It is showing the "T" relative to the `Test<T>` type rather than the "T" relative to the `ICloneable<T>` type. In other words, the pop-up is saying
> Clones a `int`.
> Returns:
> A clone of the `int`.

when it should say

> Clones a `Test<int>`.
> Returns:
> A clone of the `Test<int>`.

**Comment by ZacharyPatten** [comment]
Here are the versions the bug is reproduceable on:
- Visual Studio Community 2019 Preview 16.11.0 Preview 1.0
- Visual Studio Community 2019 16.10.0
- Visual Studio Code `ms-dotnettools.csharp` v1.23.12

Here are my `dotnet --list-sdks`:
```
1.0.4 [C:\Program Files\dotnet\sdk]
2.1.202 [C:\Program Files\dotnet\sdk]
2.1.524 [C:\Program Files\dotnet\sdk]
2.1.617 [C:\Program Files\dotnet\sdk]
2.1.700 [C:\Program Files\dotnet\sdk]
2.1.701 [C:\Program Files\dotnet\sdk]
2.1.816 [C:\Program Files\dotnet\sdk]
2.2.300 [C:\Program Files\dotnet\sdk]
2.2.301 [C:\Program Files\dotnet\sdk]
2.2.401 [C:\Program Files\dotnet\sdk]
3.1.101 [C:\Program Files\dotnet\sdk]
3.1.116 [C:\Program Files\dotnet\sdk]
5.0.100-rc.1.20452.10 [C:\Program Files\dotnet\sdk]
5.0.104 [C:\Program Files\dotnet\sdk]
5.0.204 [C:\Program Files\dotnet\sdk]
5.0.300-preview.21258.4 [C:\Program Files\dotnet\sdk]
5.0.300 [C:\Program Files\dotnet\sdk]
5.0.301 [C:\Program Files\dotnet\sdk]
6.0.100-preview.4.21255.9 [C:\Program Files\dotnet\sdk]
```

**Comment by ZacharyPatten** [comment]
@sharwell (only pinging you because I'm under the impression you are heavily involved in xml topics. sorry if you don't want pings)

**Comment by Youssef1313** [comment]
Closest thing I could get is https://github.com/Youssef1313/roslyn/commit/984c295a322d50cb01e3c8ada4714cf667a3e6ed, which *partially* fixes the issue. This changes the behavior from `int` to `Test<T>`. We want it `Test<int>`.

I haven't spent enough time on it, but thought this may help if someone will work on it. (I'm also tagging @sharwell and @CyrusNajmabadi in case the approach I'm taking isn't good/correct)

**Comment by ZacharyPatten** [comment]
This issue was not fully resolved. It should be re-opened:
![image](https://user-images.githubusercontent.com/3385986/149848319-ee76f33c-09d8-48c7-98f4-e3c543acff96.png)
_From `Visual Studio Community 2022 Preview` version `17.1.0 Preview 2.0`_

It should be showing `Test<int>` in the quick info and be a link to the type, but instead it is showing <code>Test{`0}</code>.

**Comment by ZacharyPatten** [comment]
There are also cases where generic parameters are being stripped from the types that should appear in the quick info:
![image](https://user-images.githubusercontent.com/3385986/149849192-4857be20-0f3b-4ed0-8183-7a46d1a50a3c.png)
From Visual Studio Community 2022 Preview version 17.1.0 Preview 2.0

The above example was from a class, `class BTreeLinked<T, TCompare> : ICloneable<BTreeLinked<T, TCompare>>`. The quick info on the `Clone` method should be referencing the `BTreeLinked<T, TCompare>` but instead it is referencing the type `BTreeLinked` (it is missing the generic parameters). So there are multiple different cases still affected by this issue.

_I'll share a link to the source code above example in the future once it is ready._

**Comment by Youssef1313** [comment]
@ZacharyPattern I'm not able to reproduce. Maybe my code has something different that I can't notice? 😄 

![image](https://user-images.githubusercontent.com/31348972/149897847-54763f51-a300-4301-88de-49e6c51184e3.png)

Code I used:

```csharp
internal class C
{
    public static void M()
    {
        var a = new Test<int>();
        var b = a.Clone();
    }
}
internal class Test<T> : ICloneable<Test<T>>
{
    /// <inheritdoc/>
    public Test<T> Clone() => new();
}
/// <summary>A type that has clonable instances.</summary>
/// <typeparam name="T">The type of instances that can be cloned.</typeparam>
internal interface ICloneable<T>
{
    /// <summary>Clones a <typeparamref name="T"/>.</summary>
    /// <returns>A clone of the <typeparamref name="T"/>.</returns>
    T Clone();
}
```

The behavior isn't ideal that we show `<T>` while we *should* be aware of what `T` is, but not that bad.

**Comment by ZacharyPatten** [comment]
With your code I'm still seeing the issue in my environment:
![image](https://user-images.githubusercontent.com/3385986/149942196-905179af-4951-4814-9cea-c422b480fb52.png)

Also, as mentioned when I add a non-generic class of the same name, it is displaying the non-generic class instead of the generic one (is displaying `Test` instead of `Test<int>` or `Test<T>`):
![image](https://user-images.githubusercontent.com/3385986/149942366-29059cd0-05a2-484a-ba1b-409015db844d.png)

@Youssef1313

**Comment by Youssef1313** [comment]
@ZacharyPatten Do you have any extension installed that could have overridden VS behavior? There is already a unit test for the exact same code I posted, and it also works for me locally.

**Comment by ZacharyPatten** [comment]
I reproduced my issues on 3 separate computers using either the latest 2022 or 2022 Preview with only the default extensions. And several members from the C# discord were also able to reproduce the issue. I don't have any extensions that would affect this.

**Comment by CyrusNajmabadi** [comment]
This is waht i see with your repro case on latest main:

![image](https://user-images.githubusercontent.com/4564579/150047581-204fa73c-2a13-42e1-a72b-1467e917de9d.png)

**Comment by CyrusNajmabadi** [comment]
> I reproduced my issues on 3 separate computers using either the latest 2022 or 2022 Preview with only the default extensions. And several members from the C# discord were also able to reproduce the issue.

Were you running the latest built roslyn out github.com/dotnet/roslyn main's branch though?

**Comment by ZacharyPatten** [comment]
No I was just using the latest versions of VS that have been released. If the code hasn't made it into a release yet, then I don't have it. I just saw that this was added to the `17.0.P2` milestone and assumed it should be in a release.

**Comment by Youssef1313** [comment]
I'm not really sure how this can be investigated.

@sharwell Any idea?

**Comment by Youssef1313** [comment]
@ZacharyPatten I can reproduce this in VS Code. If OmniSharp got this from Roslyn, then it may explain what's going on.

![image](https://user-images.githubusercontent.com/31348972/150116331-334a4784-f4b0-4cee-8553-5c704cad3591.png)

**Comment by sharwell** [comment]
Maybe this only reproduces with the LSP editor?

**Comment by Youssef1313** [comment]
@sharwell What is the option for that in VS?

**Comment by Youssef1313** [comment]
Ok I see this option isn't public:

https://github.com/dotnet/roslyn/blob/c5739ec1606579a43d61743c050c00f9d89787a8/src/VisualStudio/Core/Def/PackageRegistration.pkgdef#L16-L20

@ZacharyPatten Can you try running from `main` and try enabling/disabling this option and see if the bug appears and disappears as well?

**Comment by Youssef1313** [comment]
@ZacharyPatten No need to do it on your end. I confirmed it locally. Finally can reproduce:

![image](https://user-images.githubusercontent.com/31348972/150303588-708c673e-2aac-44cc-8e5e-79af7daaa48b.png)

Thanks @sharwell for pointing this out!

I'll open a new issue for it.
